dask/array/optimization.py

```diff
@@ -290,3 +299,182 @@ def fuse_slice(a, b):
+    before any actual tasks are generated, and so can reduce overhead.
+
+    This finds groups of TOP operations that can be safely fused, and then
+    passes them to ``rewrite_atop`` for rewriting.
+
+    Parameters
+    ----------
+    sharedict : ShareDict
+    keys: Iterable
+        The keys of all outputs of all collections.
+        Used to make sure that we don't fuse a layer needed by an output
+
+    Returns
+    -------
+    sharedict : ShareDict
+
+    See Also
+    --------
+    rewrite_atop
+    """
+    from .core import TOP
+    keep = {k[0] if type(k) is tuple else k for k in keys}
+    layers = sharedict.dicts
+    dependents = reverse_dict(sharedict.dependencies)
+    roots = {k for k, v in sharedict.dicts.items() if not dependents.get(k)}
+    stack = list(roots)
+
+    out = {}
+    dependencies = {}
+    seen = set()
+
+    while stack:
+        layer = stack.pop()
+        if layer in seen:
+            continue
+        seen.add(layer)
+        if isinstance(layers[layer], TOP):
+            top_layers = {layer}
+            deps = set(top_layers)
+            while deps:
+                dep = deps.pop()
+                if (isinstance(layers[dep], TOP) and
+                        not (dep != layer and dep in keep) and
+                        layers[dep].concatenate == layers[layer].concatenate):
+                    top_layers.add(dep)
+                    for d in sharedict.dependencies.get(dep, ()):
+                        if len(dependents[d]) <= 1:
+                            deps.add(d)
+                        else:
+                            stack.append(d)
+                else:
+                    stack.append(dep)
+            new_layer = rewrite_atop([layers[l] for l in top_layers])
+            out[layer] = new_layer
+            dependencies[layer] = {k for k, v in new_layer.indices if v is not None}
+        else:
+            out[layer] = layers[layer]
+            dependencies[layer] = sharedict.dependencies.get(layer, set())
+            stack.extend(sharedict.dependencies.get(layer, ()))
+
+    return ShareDict(out, dependencies)
+
+
+def rewrite_atop(inputs):
+    """ Rewrite a stack of atop expressions into a single atop expression
+
+    Given a set of TOP layers, combine them into a single layer.  The provided
+    layers are expected to fit well together.  That job is handled by
+    ``optimize_atop``
+
+    Parameters
+    ----------
+    inputs : List[TOP]
+
+    Returns
+    -------
+    top : TOP
+
+    See Also
+    --------
+    optimize_atop
+    """
+    from dask.core import reverse_dict
+    from dask.array.core import TOP
+    inputs = {inp.output: inp for inp in inputs}
+    dependencies = {inp.output: {d for d, v in inp.indices if v is not None and d in inputs}
+                    for inp in inputs.values()}
+    dependents = reverse_dict(dependencies)
+
+    new_index_iter = iter('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
```

Comment:
Any reason to use characters (which can run out) rather than a numeric suffix (e.g. `'X.1', `'X.2', ....)? I doubt we'll see 27 unique indices, but if there's no reason to limit things it'd be good to allow.